<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>GEM extraction using ftINIT - Human-GEM Guide</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "GEM extraction using ftINIT";
        var mkdocs_page_input_path = "gem_extraction.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/languages/matlab.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Human-GEM Guide
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Overview</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../getting_started/">Getting started</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../flux_balance_analysis/">Flux balance analysis</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">GEM extraction using ftINIT</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#fast-taskdriven-integrative-network-inference-for-tissues-ftinit">Fast Task‐driven Integrative Network Inference for Tissues (ftINIT)</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#retrieve-the-data">Retrieve the data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#load-and-prepare-the-reference-model-for-use-with-ftinit">Load and prepare the reference model for use with ftINIT</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#prepare-the-transcriptomic-data">Prepare the transcriptomic data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#run-ftinit">Run ftINIT</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#run-ftinit-for-all-samples">Run ftINIT for all samples</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#quick-examination-of-the-models">Quick examination of the models</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../gem_comparison/">GEM comparison</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../gem_extraction_sc/">GEM extraction from scRNA-Seq data</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../additional_tools/">Additional tools</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../faq_troubleshoot/">FAQs and troubleshooting</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../contact/">Contact</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Human-GEM Guide</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" alt="Docs"></a> &raquo;</li>
      <li>GEM extraction using ftINIT</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="gem-extraction">GEM Extraction</h1>
<p>Human-GEM is a generic model of human metabolism, meaning that it contains metabolic reactions known to occur in any human cell. The model is therefore not representative of any one tissue or cell type, in which only a subset of the reactions would be active.</p>
<p>A common approach is to extract the subset of the GEM that is likely to be active in the tissue or cell type of interest, based on a corresponding dataset (e.g., transcriptomics, proteomics, metabolomics, etc.). This approach is often referred to as <em>GEM extraction</em> or <em>contextualization</em>, where the generated model is termed the <em>extracted</em> or <em>contextualized</em>/<em>context-specific GEM</em>.</p>
<h2 id="fast-taskdriven-integrative-network-inference-for-tissues-ftinit">Fast Task‐driven Integrative Network Inference for Tissues (ftINIT)</h2>
<p>Although many GEM extraction methods exist, this guide will cover the ftINIT algorithm. For more details on the algorithm, see the following papers:</p>
<ol>
<li><a href="https://doi.org/10.1371/journal.pcbi.1002518">Agren <em>et al.</em> (2012) <em>PLoS Comput Biol</em></a> - Initial development of INIT</li>
<li><a href="https://doi.org/10.1002/msb.145122">Agren <em>et al.</em> (2014) <em>Mol Syst Biol</em></a> - Incorporation of metabolic tasks (tINIT)</li>
<li><a href="https://doi.org/10.1126/scisignal.aaz1482">Robinson <em>et al.</em> (2020) <em>Sci Signal</em></a> - Recognition of enzyme complexes</li>
<li><a href="https://doi.org/10.1101/2022.04.25.489379">Gustafsson <em>et al.</em> (2022) <em>BioRXiv</em></a> - ftINIT, faster and produces less gaps in the models</li>
</ol>
<p>The ftINIT algorithm is available in the RAVEN Toolbox, and we use it together with help functions for Human-GEM available in the Human-GEM repository.</p>
<p>To support users of the previous algorithm (tINIT), we also include a <a href="../gem_extraction_old_tINIT/">guide</a> for the previous (2020) version of tINIT, called the <code>getINITModel2</code> function.</p>
<h2 id="retrieve-the-data">Retrieve the data</h2>
<p>To demonstrate the use of ftINIT, we will walk through an example where we generate models from GTEx data. For simplicity, we extracted 60 RNA-Seq profiles from in total 12 tissues. The data can be downloaded <a href="https://doi.org/10.5281/zenodo.6811073">here</a>  together with other data useful for this tutorial. </p>
<p>Download the Zenodo repository .zip file, and extract.</p>
<h2 id="load-and-prepare-the-reference-model-for-use-with-ftinit">Load and prepare the reference model for use with ftINIT</h2>
<p>The reference GEM from which the tissue-specific models will be extracted is Human-GEM. Load the model from the <code>Human-GEM.mat</code> file in the Human-GEM repository</p>
<pre><code class="language-matlab">load('Human-GEM.mat');  % loads model as a structure named &quot;ihuman&quot;
</code></pre>
<p>ftINIT has a preparation step that needs to be run once for a reference model such as Human-GEM. The purpose of the preparation step is to make calculations in advance to reduce the time it takes to generate each context-specific model. The result is a prepData structure that contains various information needed by ftINIT. This operation takes 1-2 hours on a standard laptop computer, so make sure to save it once it has completed. For version 1.12.0 of Human-GEM, the prepData is available in the Zenodo .zip file.</p>
<pre><code class="language-matlab">% The second flag indicates if the model should be converted to gene symbols from ENSEMBL. This has to be decided at this point.
prepData = prepHumanModelForftINIT(ihuman, false);
save('prepData.mat', 'prepData')
</code></pre>
<p><code>prepHumanModelForftINIT</code> tests that the reference model (Human-GEM) can successfully perform all essential metabolic tasks (defined in <code>data/metabolicTasks/metabolicTasks_Essential.txt</code>), which is an important step. If the reference model cannot perform a task, then neither can any GEM extracted from that model.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The exchange reactions should all be unbounded (e.g., lower and upper bounds of -1000 and 1000, respectively). Human-GEM is already provided in this format, so no additional changes are needed.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The function <code>prepHumanModelForftINIT</code> will automatically use the essential tasks from <code>metabolicTasks_Essential.txt</code> in the Human-GEM repository. To specify different metabolic tasks, use <code>prepINITModel</code> in RAVEN directly instead.</p>
</div>
<h2 id="prepare-the-transcriptomic-data">Prepare the transcriptomic data</h2>
<p>Data from the condition, tissue, or cell type for which the GEM will be generated is necessary to determine which reactions from the reference model (Human-GEM) should be included. In this case, we use a file with tissues extracted from the GTEx database, which is available in the Zenodo .zip file.</p>
<p>In MATLAB, load the <code>gtexSampForTutorialTPM.txt</code> file from the Zenodo repository.</p>
<pre><code class="language-matlab">% replace 'my/path/' with the path on your system, or change to the directory containing the file
gtex_data = readtable('my/path/gtexSampForTutorialTPM.txt');
[~, n] = size(gtex_data);
numSamp = n-2; %the first two columns are the genes in ENSEMBL and gene symbols format

% take a look at the first few rows and columns of the table
gtex_data(1:5, 1:5)
% ans =
% 
%   5×5 table
% 
%            Name              Description      GTEX_111YS_0006_SM_5NQBE    GTEX_1122O_0005_SM_5O99J    GTEX_1128S_0005_SM_5P9HI
%     ___________________    _______________    ________________________    ________________________    ________________________
% 
%     {'ENSG00000223972'}    {'DDX11L1'    }            0.02171                     0.03015                      0.0203         
%     {'ENSG00000227232'}    {'WASH7P'     }              1.555                       2.747                       4.441         
%     {'ENSG00000278267'}    {'MIR6859-1'  }                  0                           0                           0         
%     {'ENSG00000243485'}    {'MIR1302-2HG'}                  0                           0                           0         
%     {'ENSG00000237613'}    {'FAM138A'    }                  0                           0                      0.0576         
</code></pre>
<p>Extract information from the table into a structure called <code>data_struct</code></p>
<pre><code class="language-matlab">% extract the tissue and gene names
data_struct.genes = gtex_data{:, 1}; % gene names
data_struct.tissues = gtex_data.Properties.VariableNames(3:n); % sample (tissue) names
data_struct.levels = gtex_data{:, 3:n}; % gene TPM values
</code></pre>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The name of the data structure (<code>data_struct</code> in this example) is not important. However, its fields (<code>genes</code>, <code>tissues</code>, <code>levels</code>) should <strong>not</strong> be changed.</p>
</div>
<p>Although the original tINIT implementation always compared gene expression in the tissue of interest to the average of all other provided tissues, ftINIT enables the option to instead compare the expression to a threshold value. Here, we will use this alternative approach with a threshold value of 1 TPM.</p>
<pre><code class="language-matlab">data_struct.threshold = 1;
</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Exclude the <code>threshold</code> field from the data structure to use the original tINIT approach of comparing the gene expression in your tissue of interest with all other tissues. If you choose this approach, however, you <em><strong>must include the gene expression data from the other tissues</strong></em> so that the algorithm has something to compare to.</p>
</div>
<p>Take a look at <code>data_struct</code> to make sure all the fields are present and have the expected dimensions</p>
<pre><code class="language-matlab">data_struct

% data_struct =
%
%   struct with fields:
% 
%         genes: {56200×1 cell}
%       tissues: {1×60 cell}
%        levels: [56200×60 double]
%     threshold: 1
</code></pre>
<h2 id="run-ftinit">Run ftINIT</h2>
<p>Now all inputs are ready to run ftINIT and extract GEMs specific to the samples based on their corresponding RNA expression profile. ftINIT normally runs in two steps, of which the second is optional. The first step excludes most of the reactions without gene rules (GPRs) from the problem, and the second step determines which of those reactions should be removed. The second step can be omitted, which causes most reactions without GPRs to remain in the model. This is a good option in many cases, for example for structural comparison of models, since removal of reactions without GPRs does not provide any additional information and may add randomness in cases where there are several equally good solutions.</p>
<p>We first run ftINIT without the second step, a setup that is called <code>'1+0'</code>, which typically takes 30 - 60 seconds:</p>
<pre><code class="language-matlab">model1 = ftINIT(prepData, data_struct.tissues{1}, [], [], data_struct, {}, getHumanGEMINITSteps('1+0'), false, true);

model1
% 
%   struct with fields:
% 
%                      id: 'Human-GEM'
%             description: 'Generic genome-scale metabolic model of Homo sapiens'
%                    rxns: {8407×1 cell}
%                    mets: {5613×1 cell}
%                       S: [5613×8407 double]
%                      lb: [8407×1 double]
%                      ub: [8407×1 double]
%                     rev: [8407×1 double]
%                       c: [8407×1 double]
%                       b: [5613×1 double]
%                   comps: {9×1 cell}
%               compNames: {9×1 cell}
%                rxnNames: {8407×1 cell}
%                 grRules: {8407×1 cell}
%              rxnGeneMat: [8407×2494 double]
%              subSystems: {8407×1 cell}
%                 eccodes: {8407×1 cell}
%                rxnNotes: {8407×1 cell}
%                   genes: {2494×1 cell}
%                metNames: {5613×1 cell}
%                metComps: [5613×1 double]
%                  inchis: {5613×1 cell}
%             metFormulas: {5613×1 cell}
%           rxnReferences: {8407×1 cell}
%                 rxnFrom: {8407×1 cell}
%                 metFrom: {5613×1 cell}
%     rxnConfidenceScores: [8407×1 double]
%              metCharges: [5613×1 int64]
%                 version: '1.12.0'
%              annotation: [1×1 struct]
</code></pre>
<p>As an alternative, we can run it with the second step included (<code>'1+1'</code>), which takes roughly 2 - 3 times as long and generates a smaller model:</p>
<pre><code class="language-matlab">model2 = ftINIT(prepData, data_struct.tissues{1}, [], [], data_struct, {}, getHumanGEMINITSteps('1+1'), false, true);

model2
% 
%   struct with fields:
% 
%                      id: 'Human-GEM'
%             description: 'Generic genome-scale metabolic model of Homo sapiens'
%                    rxns: {7752×1 cell}
%                    mets: {5490×1 cell}
%                       S: [5490×7752 double]
%                      lb: [7752×1 double]
%                      ub: [7752×1 double]
%                     rev: [7752×1 double]
%                       c: [7752×1 double]
%                       b: [5490×1 double]
%                   comps: {9×1 cell}
%               compNames: {9×1 cell}
%                rxnNames: {7752×1 cell}
%                 grRules: {7752×1 cell}
%              rxnGeneMat: [7752×2494 double]
%              subSystems: {7752×1 cell}
%                 eccodes: {7752×1 cell}
%                rxnNotes: {7752×1 cell}
%                   genes: {2494×1 cell}
%                metNames: {5490×1 cell}
%                metComps: [5490×1 double]
%                  inchis: {5490×1 cell}
%             metFormulas: {5490×1 cell}
%           rxnReferences: {7752×1 cell}
%                 rxnFrom: {7752×1 cell}
%                 metFrom: {5490×1 cell}
%     rxnConfidenceScores: [7752×1 double]
%              metCharges: [5490×1 int64]
%                 version: '1.12.0'
%              annotation: [1×1 struct]
</code></pre>
<p>It is also possible to supply cell type for cases where tissues are subdivided into cell type, which is not the case here. The method also accepts proteomics data from Human Protein Atlas (HPA) and metabolomics data, but these are not demonstrated here.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The two steps used here are the two most common approaches to run ftINIT and are what we recommend, but they are not the only options. ftINIT can for example be run in a very similar way to the original tINIT algorithm by using the <code>'full'</code> setup.</p>
</div>
<p>It is recommended to change the model <code>id</code> to a more descriptive name than the default of "INITModel". This is particularly useful when analyzing several models together.</p>
<pre><code class="language-matlab">model1.id = data_struct.tissues{1};
</code></pre>
<h2 id="run-ftinit-for-all-samples">Run ftINIT for all samples</h2>
<p>We can now run ftINIT on all samples:</p>
<pre><code class="language-matlab">models = cell(numSamp, 1);
for i = 1:numSamp
    disp(['Model: ' num2str(i) ' of ' num2str(numSamp)])
    models{i} = ftINIT(prepData, data_struct.tissues{i}, [], [], data_struct, {}, getHumanGEMINITSteps('1+0'), false, true);
end

save('models.mat', 'models')
</code></pre>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This loop generates 60 models, which will take a while to run (30 min to 1 hr). For running many models, it is therefore recommended that this is run on a compute cluster.</p>
</div>
<h2 id="quick-examination-of-the-models">Quick examination of the models</h2>
<p>To get a quick structural overview of the models, we perform a t-SNE (t-Distributed Stochastic Neighbor Embedding) transformation of the models' reaction content and export that data to a file for visualization in R.</p>
<pre><code class="language-matlab">baseModel = prepData.refModel;

% now build a matrix saying which reactions are on
compMat = false(length(baseModel.rxns), length(models));

for i = 1:size(compMat,2)
    compMat(:,i) = ismember(baseModel.rxns,models{i}.rxns);
end

% run t-sne
rng(1);  %set random seed to make reproducible
proj_coords = tsne(double(compMat.'), 'Distance', 'hamming', 'NumDimensions', 2, 'Exaggeration', 6, 'Perplexity', 10);

% export to R
d = struct();
d.tsneX = proj_coords(:, 1);
d.tsneY = proj_coords(:, 2);
save('TSNE.mat', 'd');
</code></pre>
<p>We can then visualize the data in R (use e.g., RStudio) using ggplot:</p>
<pre><code class="language-R">library(R.matlab)
library(tidyverse)
library(ggplot2)

#import the data using R.matlab
setwd(&quot;my/path&quot;) #replace with your own path where the exported data was saved
tsneData = readMat(&quot;TSNE.mat&quot;)
x = as.numeric(tsneData$d[,,1]$tsneX)
y = as.numeric(tsneData$d[,,1]$tsneY)

#read the tissues to be able to color the models properly
tissues = read_tsv('gtexSampTissuesForTutorial.txt', col_names = FALSE)[[1]]
tissFact = as.factor(tissues)

#set up a tibble for the plot
df = tibble(x=x, y=y, tissue = tissFact)

color_palette = c('#B5D39B','#E7B56C','#6B97BC','#BC976B','#BC556B','#000000')  

#plot the data
fig = ggplot(df, aes(x = x, y = y, color=tissue, shape=tissue)) +
  geom_point(size=2, stroke = 2) +
  scale_color_manual(values = c(color_palette,color_palette), labels = levels(tissFact) ) +
  scale_shape_manual(values = c(19,19,19,19,19,19,0,0,0,0,0,0), labels=levels(tissFact)) +
  ggplot2::labs(y=expression(&quot;t-SNE y&quot;), x=&quot;t-SNE x&quot;, title=&quot;Structural comparison&quot;) +
  ggplot2::theme_bw() + 
  ggplot2::theme(panel.background = element_rect(&quot;white&quot;, &quot;white&quot;, 0, 0, &quot;white&quot;), panel.grid.major= element_blank(),panel.grid.minor= element_blank()) +
  ggplot2::theme(legend.title = element_blank(),legend.position=&quot;bottom&quot;, legend.text=element_text(size=14)) + guides(colour = guide_legend(nrow = 4), size = guide_legend(nrow = 4), linetype = guide_legend(nrow = 4)) +
  ggplot2::theme(text = element_text(size=14),
                 axis.text.x = element_text(color='black', size=14),
                 axis.text.y = element_text(color='black', size=14))
fig

#export the figure to file
ggsave(
  &quot;StructCompftINIT.png&quot;,
  plot = fig,
  width = 8, height = 8, dpi = 300)
</code></pre>
<p><img alt="Structural comparison" src="../img/StructCompftINIT.png" style="width:95%" /></p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../flux_balance_analysis/" class="btn btn-neutral float-left" title="Flux balance analysis"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../gem_comparison/" class="btn btn-neutral float-right" title="GEM comparison">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../flux_balance_analysis/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../gem_comparison/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
