<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>GEM extraction using tINIT - Human-GEM Guide</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "GEM extraction using tINIT";
    var mkdocs_page_input_path = "gem_extraction.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/matlab.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Human-GEM Guide</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Overview</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../getting_started/">Getting started</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../flux_balance_analysis/">Flux balance analysis</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">GEM extraction using tINIT</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#taskdriven-integrative-network-inference-for-tissues-tinit">Task‐driven Integrative Network Inference for Tissues (tINIT)</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#example-extract-a-liver-gem">Example: Extract a liver GEM</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#prepare-the-transcriptomic-data">Prepare the transcriptomic data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#load-the-reference-gem">Load the reference GEM</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#prepare-the-metabolic-task-list">Prepare the metabolic task list</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#run-tinit">Run tINIT</a>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../gem_comparison/">GEM comparison</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Human-GEM Guide</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>GEM extraction using tINIT</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="gem-extraction">GEM Extraction</h1>
<p>Human-GEM is a generic model of human metabolism, meaning that it contains metabolic reactions known to occur in any human cell. The model is therefore not representative of any one tissue or cell type, in which only a subset of the reactions would be active.</p>
<p>A common approach is to extract the subset of the GEM that is likely to be active in the tissue or cell type of interest, based on a corresponding dataset (e.g., transcriptomics, proteomics, metabolomics, etc.). This approach is often referred to as <em>GEM extraction</em> or <em>contextualization</em>, where the generated model is termed the <em>extracted</em> or <em>contextualized</em>/<em>context-specific GEM</em>.</p>
<h2 id="taskdriven-integrative-network-inference-for-tissues-tinit">Task‐driven Integrative Network Inference for Tissues (tINIT)</h2>
<p>Although many GEM extraction methods exist, this guide will cover the tINIT algorithm. For more details on the algorithm itself, see the following papers:</p>
<ol>
<li><a href="https://doi.org/10.1371/journal.pcbi.1002518">Agren <em>et al.</em> (2012) <em>PLoS Comput Biol</em></a> - Initial development of INIT</li>
<li><a href="https://doi.org/10.1002/msb.145122">Agren <em>et al.</em> (2014) <em>Mol Syst Biol</em></a> - Incorporation of metabolic tasks (tINIT)</li>
<li><a href="https://doi.org/10.1126/scisignal.aaz1482">Robinson <em>et al.</em> (2020) <em>Sci Signal</em></a> - Recognition of enzyme complexes</li>
</ol>
<p>The original (2014) tINIT algorithm is implemented in the RAVEN Toolbox, where its main calling function is <code>getINITModel</code>. This guide will cover the updated (2020) version of tINIT that is maintained on the Human-GEM repository, called using the <code>getINITModel2</code> function.</p>
<h2 id="example-extract-a-liver-gem">Example: Extract a liver GEM</h2>
<p>To demonstrate the use of tINIT, we will walk through an example that replicates part of the <a href="https://doi.org/10.1126/scisignal.aaz1482">Human-GEM publication</a>. </p>
<p>The data used in this example can be retrieved from the <a href="https://doi.org/10.5281/zenodo.3577466">Zenodo repository</a> associated with the publication. Download the Zenodo repository .zip file, and extract.</p>
<h4 id="prepare-the-transcriptomic-data">Prepare the transcriptomic data</h4>
<p>Data from the condition, tissue, or cell type for which the GEM will be generated is necessary to determine which reactions from the reference model (Human-GEM) should be included. In this case, we are using RNA-Seq data (TPM normalized) that was originally obtained from the <a href="https://gtexportal.org/home/">GTEx database</a>. </p>
<p>In MATLAB, load the <code>gtex_median_tissue_tpm.txt</code> file from the Zenodo repository <code>tINIT_GEMs/data/</code> subdirectory.</p>
<pre><code class="matlab">% replace 'my/path/' with the path on your system, or change to the directory with the file
gtex_data = readtable('my/path/tINIT_GEMs/data/gtex_median_tissue_tpm.txt');

% take a look at the first few rows and columns of the table
gtex_data(1:5,1:5)

% ans =
% 
%   5×5 table
% 
%            genes           adipose_tissue    adrenal_gland    blood_vessel    bladder
%     ___________________    ______________    _____________    ____________    _______
% 
%     {'ENSG00000223972'}       0.05458            0.0746         0.04535       0.05878
%     {'ENSG00000227232'}          11.2             8.023           11.99         14.24
%     {'ENSG00000243485'}       0.06012           0.08179         0.05278       0.06097
%     {'ENSG00000237613'}       0.03653            0.0405         0.03668       0.04113
%     {'ENSG00000268020'}       0.02565           0.03479               0             0
</code></pre>

<p>Extract information from the table into a structure called <code>data_struct</code></p>
<pre><code class="matlab">% extract the tissue and gene names
data_struct.tissues = gtex_data.Properties.VariableNames(2:end)';  % sample (tissue) names
data_struct.genes = gtex_data.genes;  % gene names
data_struct.levels = table2array(gtex_data(:, 2:end));  % gene TPM values
</code></pre>

<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The name of the data structure (<code>data_struct</code> in this example) is not important. However, its fields (<code>genes</code>, <code>tissues</code>, <code>levels</code>) should <strong>not</strong> be changed.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We don't actually need all of the data for this example since we are only generating a model for one tissue (liver). One could instead extract only the <code>liver</code> column into the <code>levels</code> field, and included only <code>liver</code> in the <code>tissues</code> field.</p>
</div>
<p>Although the original tINIT implementation always compared gene expression in the tissue of interest to the average of all other provided tissues, the updated version enables the option to instead compare the expression to a threshold value. Here, we will use this alternative approach with a threshold value of 1 TPM.</p>
<pre><code class="matlab">data_struct.threshold = 1;
</code></pre>

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Exclude the <code>threshold</code> field from the data structure to use the original tINIT approach of comparing the gene expression in your tissue of interest with all other tissues. If you choose this approach, however, you <em><strong>must include the gene expression data from the other tissues</strong></em> so that the algorithm has something to compare to.</p>
</div>
<p>Take a look at <code>data_struct</code> to make sure all the fields are present and have the expected dimensions</p>
<pre><code class="matlab">data_struct

% data_struct =
%
%   struct with fields:
% 
%       tissues: {30×1 cell}
%         genes: {56202×1 cell}
%        levels: [56202×30 double]
%     threshold: 1
</code></pre>

<h4 id="load-the-reference-gem">Load the reference GEM</h4>
<p>The reference GEM from which the liver-specific model will be extracted is Human-GEM. Load the model from the <code>HumanGEM.mat</code> file in the Human-GEM repository</p>
<pre><code class="matlab">load('HumanGEM.mat');  % loads model as a structure named &quot;ihuman&quot;
</code></pre>

<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The tINIT algorithm requires that the model is provided in "closed form"; i.e., it contains boundary metabolites. Therefore, the boundary metabolites should <strong>not</strong> be removed by e.g., the <code>simplifyModel</code> function before use with tINIT. At the same time, the exchange reactions should all be unbounded (lower and upper bounds of -1000 and 1000, respectively). Human-GEM is already provided in a format that satisfies both of these requirements, so no additional changes are needed.</p>
</div>
<h4 id="prepare-the-metabolic-task-list">Prepare the metabolic task list</h4>
<p>The tINIT algorithm has the option to specify one or more metabolic tasks that the extracted model should be able to perform, such as transport of essential amino acids or production of biomass. Although task specification is not required, it is generally recommended to obtain a more functional extracted GEM, especially if it is going to be used in simulation-based analysis (e.g., FBA).</p>
<p>Following with the procedure used in the Human-GEM manuscript, this example will use the list of 57 "essential" metabolic tasks that all cells are expected to be able to perform to be viable.</p>
<p>Load the <code>metabolicTasks_Essential.xlsx</code> list of metabolic tasks from the <code>data/metabolicTasks/</code> subdirectory of the Human-GEM repository using the RAVEN <code>parseTaskList</code> function.</p>
<pre><code class="matlab">essentialTasks = parseTaskList('/my/path/Human-GEM/data/metabolicTasks/metabolicTasks_Essential.xlsx')

% essentialTasks = 
% 
%   57×1 struct array with fields:
% 
%     id
%     description
%     shouldFail
%     printFluxes
%     comments
%     inputs
%     LBin
%     UBin
%     outputs
%     LBout
%     UBout
%     equations
%     LBequ
%     UBequ
%     changed
%     LBrxn
%     UBrxn
</code></pre>

<p>It is important to first verify that the reference model (Human-GEM) can successfully perform all of the tasks. If the reference model cannot perform a task, then neither can any GEM extracted from that model.</p>
<p>Verify that the reference model passes all of the essential metabolic tasks.</p>
<pre><code class="matlab">% see what the other inputs mean by typing &quot;help checkTasks&quot;
checkTasks(ihuman, [], true, false, false, essentialTasks);

% PASS: [ER] Aerobic rephosphorylation of ATP from glucose
% PASS: [ER] Aerobic rephosphorylation of GTP
% PASS: [ER] Aerobic rephosphorylation of CTP
% PASS: [ER] Aerobic rephosphorylation of UTP
% PASS: [BS] ATP de novo synthesis
% PASS: [BS] CTP de novo synthesis
% PASS: [BS] GTP de novo synthesis
% PASS: [BS] UTP de novo synthesis
% PASS: [BS] dATP de novo synthesis
% ...
</code></pre>

<h4 id="run-tinit">Run tINIT</h4>
<p>Now all inputs are ready to run tINIT and extract a GEM specific to the liver tissue based on its corresponding RNA expression profile. </p>
<p>Although it's not necessary to re-assign the input variables before calling the <code>getINITModel2</code> function, we do it here to make the inputs very clear. Use the <code>help getINITModel2</code> command to see more details on the function inputs and outputs.</p>
<pre><code class="matlab">refModel = ihuman;  % the reference model from which the GEM will be extracted
tissue = 'liver';  % must match the tissue name in data_struct.tissues
celltype = [];  % used if tissues are subdivided into cell type, which is not the case here
hpaData = [];  % data structure containing protein abundance information (not used here)
arrayData = data_struct;  % data structure with gene (RNA) abundance information
metabolomicsData = [];  % list of metabolite names if metabolomic data is available
removeGenes = true;  % (default) remove lowly/non-expressed genes from the extracted GEM
taskFile = [];  % we already loaded the task file, so this input is not required
useScoresForTasks = true;  % (default) use expression data to decide which reactions to keep
printReport = true;  % (default) print status/completion report to screen
taskStructure = essentialTasks;  % metabolic task structure (used instead &quot;taskFile&quot;)
params = [];  % additional optimization parameters for the INIT algorithm
paramsFT = [];  % additional optimization parameters for the fit-tasks algorithm
</code></pre>

<p>Run the getINITModel2 function to extract the GEM</p>
<pre><code class="matlab">liverGEM = getINITModel2(refModel, tissue, celltype, hpaData, arrayData, metabolomicsData, removeGenes, taskFile, useScoresForTasks, printReport, taskStructure, params, paramsFT);

% ***Generating model for: liver
% -Using array data
% -Using metabolic tasks
% 
% Reference model statistics:
%         13417 reactions, 3625 genes
%         Mean reaction score: 4.1017
%         Mean gene score: 5.3912
%         Reactions with positive scores: 54.2372%
%
%  Pruned model statistics:
%         11903 reactions, 3052 genes
%         Mean reaction score: 4.1094
%         Mean gene score: 5.1854
%         Reactions with positive scores: 54.1964%
%  
%  Reactions essential for tasks:
%         403 reactions, 365 genes
%         Mean reaction score: 8.3307
%         Mean gene score: 7.1855
%         Reactions with positive scores: 87.8412%

MILP detected.
Academic license - for non-commercial use only
Gurobi Optimizer version 9.0.1 build v9.0.1rc0 (mac64)
Optimize a model with 25830 rows, 60561 columns and 126189 nonzeros
...
</code></pre>

<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This algorithm will take a while to run (30 min to 1 hr). It is therefore recommended that this is run on a compute cluster, especially when repeating for many different GEMs.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Unfortunately, there is currently a bug in how RAVEN processes the optimization parameters passed by the <code>params</code> and <code>paramsFT</code> arguments when using the Gurobi solver. Therefore, optimization parameters (such as time limit) need to be changed manually within the RAVEN <code>optimizeProb</code> function.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In this example, we called <code>getINITModel2</code> with one output, <code>liverGEM</code>. Use <code>help getINITModel2</code> to see a description of the other optional outputs.</p>
</div>
<p>Once the optimization is complete, we can take a look at the extracted GEM.</p>
<pre><code class="matlab">liverGEM

% liverGEM = 
% 
%   struct with fields:
% 
%                      id: 'INITModel'
%             description: 'Automatically generated model for liver'
%                    rxns: {7971×1 cell}
%                    mets: {6744×1 cell}
%                       S: [6744×7971 double]
%                      lb: [7971×1 double]
%                      ub: [7971×1 double]
%                     rev: [7971×1 double]
%                       c: [7971×1 double]
%                       b: [6744×2 double]
%                   comps: {10×1 cell}
%               compNames: {10×1 cell}
%                rxnNames: {7971×1 cell}
%                 grRules: {7971×1 cell}
%              rxnGeneMat: [7971×2305 double]
%              subSystems: {7971×1 cell}
%                 eccodes: {7971×1 cell}
%                   genes: {2305×1 cell}
%                metNames: {6744×1 cell}
%                metComps: [6744×1 double]
%                  inchis: {6744×1 cell}
%             metFormulas: {6744×1 cell}
%           unconstrained: [6744×1 double]
%           rxnReferences: {7971×1 cell}
%                 rxnFrom: {7971×1 cell}
%                 metFrom: {6744×1 cell}
%     rxnConfidenceScores: [7971×1 double]
%              metCharges: [6744×1 int64]
%                 version: '1.3.1'
%              annotation: [1×1 struct]
%                geneFrom: {2305×1 cell}
</code></pre>

<p>Notice that the model has fewer reactions, metabolites, and genes compared to the reference model.</p>
<pre><code class="matlab">numel(ihuman.rxns) - numel(liverGEM.rxns)

% ans =
% 
%         5446

numel(ihuman.mets) - numel(liverGEM.mets)

% ans =
% 
%         3394

numel(ihuman.genes) - numel(liverGEM.genes)

% ans =
% 
%         1320
</code></pre>

<p>It is also good to verify that the extracted GEM can also perform the required metabolic tasks.</p>
<pre><code class="matlab">checkTasks(liverGEM, [], true, false, false, essentialTasks);
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../gem_comparison/" class="btn btn-neutral float-right" title="GEM comparison">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../flux_balance_analysis/" class="btn btn-neutral" title="Flux balance analysis"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../flux_balance_analysis/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../gem_comparison/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
