
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.1.0">
    
    
      
        <title>GEM Extraction with tINIT (the older algorithm for extracting context-specific models) - Human-GEM Guide</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.402914a4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.a0c5b2b5.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent="">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gem-extraction-with-tinit-the-older-algorithm-for-extracting-context-specific-models" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Human-GEM Guide" class="md-header__button md-logo" aria-label="Human-GEM Guide" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Human-GEM Guide
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              GEM Extraction with tINIT (the older algorithm for extracting context-specific models)
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Human-GEM Guide" class="md-nav__button md-logo" aria-label="Human-GEM Guide" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Human-GEM Guide
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Overview
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        Installation
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../getting_started/" class="md-nav__link">
        Getting started
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../flux_balance_analysis/" class="md-nav__link">
        Flux balance analysis
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../gem_extraction/" class="md-nav__link">
        GEM extraction using ftINIT
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../gem_comparison/" class="md-nav__link">
        GEM comparison
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../gem_extraction_sc/" class="md-nav__link">
        GEM extraction from scRNA-Seq data
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../additional_tools/" class="md-nav__link">
        Additional tools
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../faq_troubleshoot/" class="md-nav__link">
        FAQs and troubleshooting
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../contact/" class="md-nav__link">
        Contact
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#retrieve-the-data" class="md-nav__link">
    Retrieve the data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prepare-the-transcriptomic-data" class="md-nav__link">
    Prepare the transcriptomic data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#load-and-prepare-the-reference-gem" class="md-nav__link">
    Load and prepare the reference GEM
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prepare-the-metabolic-task-list" class="md-nav__link">
    Prepare the metabolic task list
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-tinit" class="md-nav__link">
    Run tINIT
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="gem-extraction-with-tinit-the-older-algorithm-for-extracting-context-specific-models">GEM Extraction with tINIT (the older algorithm for extracting context-specific models)</h1>
<p>The original (2014) tINIT algorithm is implemented in the RAVEN Toolbox, where its main calling function is <code>getINITModel</code>. This guide is for the updated (2020) version of tINIT that is maintained on the Human-GEM repository, called the <code>getINITModel2</code> function.</p>
<h2 id="retrieve-the-data">Retrieve the data</h2>
<p>To demonstrate the use of tINIT, we will walk through an example that replicates part of the <a href="https://doi.org/10.1126/scisignal.aaz1482">Human-GEM publication</a>. </p>
<p>The data used in this example can be retrieved from the <a href="https://doi.org/10.5281/zenodo.3577466">Zenodo repository</a> associated with the publication. Download the Zenodo repository .zip file, and extract.</p>
<h2 id="prepare-the-transcriptomic-data">Prepare the transcriptomic data</h2>
<p>Data from the condition, tissue, or cell type for which the GEM will be generated is necessary to determine which reactions from the reference model (Human-GEM) should be included. In this case, we are using RNA-Seq data (TPM normalized) that was originally obtained from the <a href="https://gtexportal.org/home/">GTEx database</a>. </p>
<p>In MATLAB, load the <code>gtex_median_tissue_tpm.txt</code> file from the Zenodo repository <code>tINIT_GEMs/data/</code> subdirectory.</p>
<pre><code class="language-matlab">% replace 'my/path/' with the path on your system, or change to the directory with the file
gtex_data = readtable('my/path/tINIT_GEMs/data/gtex_median_tissue_tpm.txt');

% take a look at the first few rows and columns of the table
gtex_data(1:5,1:5)

% ans =
% 
%   5×5 table
% 
%            genes           adipose_tissue    adrenal_gland    blood_vessel    bladder
%     ___________________    ______________    _____________    ____________    _______
% 
%     {'ENSG00000223972'}       0.05458            0.0746         0.04535       0.05878
%     {'ENSG00000227232'}          11.2             8.023           11.99         14.24
%     {'ENSG00000243485'}       0.06012           0.08179         0.05278       0.06097
%     {'ENSG00000237613'}       0.03653            0.0405         0.03668       0.04113
%     {'ENSG00000268020'}       0.02565           0.03479               0             0
</code></pre>
<p>Extract information from the table into a structure called <code>data_struct</code></p>
<pre><code class="language-matlab">% extract the tissue and gene names
data_struct.tissues = gtex_data.Properties.VariableNames(2:end)';  % sample (tissue) names
data_struct.genes = gtex_data.genes;  % gene names
data_struct.levels = table2array(gtex_data(:, 2:end));  % gene TPM values
</code></pre>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The name of the data structure (<code>data_struct</code> in this example) is not important. However, its fields (<code>genes</code>, <code>tissues</code>, <code>levels</code>) should <strong>not</strong> be changed.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We don't actually need all of the data for this example since we are only generating a model for one tissue (liver). One could instead extract only the <code>liver</code> column into the <code>levels</code> field, and included only <code>liver</code> in the <code>tissues</code> field.</p>
</div>
<p>Although the original tINIT implementation always compared gene expression in the tissue of interest to the average of all other provided tissues, the updated version enables the option to instead compare the expression to a threshold value. Here, we will use this alternative approach with a threshold value of 1 TPM.</p>
<pre><code class="language-matlab">data_struct.threshold = 1;
</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Exclude the <code>threshold</code> field from the data structure to use the original tINIT approach of comparing the gene expression in your tissue of interest with all other tissues. If you choose this approach, however, you <em><strong>must include the gene expression data from the other tissues</strong></em> so that the algorithm has something to compare to.</p>
</div>
<p>Take a look at <code>data_struct</code> to make sure all the fields are present and have the expected dimensions</p>
<pre><code class="language-matlab">data_struct

% data_struct =
%
%   struct with fields:
% 
%       tissues: {30×1 cell}
%         genes: {56202×1 cell}
%        levels: [56202×30 double]
%     threshold: 1
</code></pre>
<h2 id="load-and-prepare-the-reference-gem">Load and prepare the reference GEM</h2>
<p>The reference GEM from which the liver-specific model will be extracted is Human-GEM. Load the model from the <code>Human-GEM.mat</code> file in the Human-GEM repository</p>
<pre><code class="language-matlab">load('Human-GEM.mat');  % loads model as a structure named &quot;ihuman&quot;
</code></pre>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The metabolic task evaluation and tINIT algorithms requires that the model is provided in "closed form"; i.e., it contains <strong>boundary metabolites</strong>. Therefore, the boundary metabolites must be added to the GEM before proceeding.</p>
</div>
<p>Add boundary metabolites to the model</p>
<pre><code class="language-matlab">ihuman = addBoundaryMets(ihuman);

% Boundary metabolites were added to 1665 reactions.
% New (boundary) versions of 1665 metabolites were added to the model.
</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The exchange reactions should all be unbounded (e.g., lower and upper bounds of -1000 and 1000, respectively). Human-GEM is already provided in this format, so no additional changes are needed.</p>
</div>
<h2 id="prepare-the-metabolic-task-list">Prepare the metabolic task list</h2>
<p>The tINIT algorithm has the option to specify one or more metabolic tasks that the extracted model should be able to perform, such as transport of essential amino acids or production of biomass. Although task specification is not required, it is generally recommended to obtain a more functional extracted GEM, especially if it is going to be used in simulation-based analysis (e.g., FBA).</p>
<p>Following with the procedure used in the Human-GEM manuscript, this example will use the list of 57 "essential" metabolic tasks that all cells are expected to be able to perform to be viable.</p>
<p>Load the <code>metabolicTasks_Essential.txt</code> list of metabolic tasks from the <code>data/metabolicTasks/</code> subdirectory of the Human-GEM repository using the RAVEN <code>parseTaskList</code> function.</p>
<pre><code class="language-matlab">essentialTasks = parseTaskList('/my/path/Human-GEM/data/metabolicTasks/metabolicTasks_Essential.txt')

% essentialTasks = 
% 
%   57×1 struct array with fields:
% 
%     id
%     description
%     shouldFail
%     printFluxes
%     comments
%     inputs
%     LBin
%     UBin
%     outputs
%     LBout
%     UBout
%     equations
%     LBequ
%     UBequ
%     changed
%     LBrxn
%     UBrxn
</code></pre>
<p>It is important to first verify that the reference model (Human-GEM) can successfully perform all of the tasks. If the reference model cannot perform a task, then neither can any GEM extracted from that model.</p>
<p>Verify that the reference model passes all of the essential metabolic tasks.</p>
<pre><code class="language-matlab">% see what the other inputs mean by typing &quot;help checkTasks&quot;
checkTasks(ihuman, [], true, false, false, essentialTasks);

% PASS: [ER] Aerobic rephosphorylation of ATP from glucose
% PASS: [ER] Aerobic rephosphorylation of GTP
% PASS: [ER] Aerobic rephosphorylation of CTP
% PASS: [ER] Aerobic rephosphorylation of UTP
% PASS: [BS] ATP de novo synthesis
% PASS: [BS] CTP de novo synthesis
% PASS: [BS] GTP de novo synthesis
% PASS: [BS] UTP de novo synthesis
% PASS: [BS] dATP de novo synthesis
% ...
</code></pre>
<h2 id="run-tinit">Run tINIT</h2>
<p>Now all inputs are ready to run tINIT and extract a GEM specific to the liver tissue based on its corresponding RNA expression profile. </p>
<p>Although it's not necessary to re-assign the input variables before calling the <code>getINITModel2</code> function, we do it here to make the inputs very clear. Use the <code>help getINITModel2</code> command to see more details on the function inputs and outputs.</p>
<pre><code class="language-matlab">refModel = ihuman;          % the reference model from which the GEM will be extracted
tissue = 'liver';           % must match the tissue name in data_struct.tissues
celltype = [];              % used if tissues are subdivided into cell type, which is not the case here
hpaData = [];               % data structure containing protein abundance information (not used here)
arrayData = data_struct;    % data structure with gene (RNA) abundance information
metabolomicsData = [];      % list of metabolite names if metabolomic data is available
removeGenes = true;         % (default) remove lowly/non-expressed genes from the extracted GEM
taskFile = [];              % we already loaded the task file, so this input is not required
useScoresForTasks = true;   % (default) use expression data to decide which reactions to keep
printReport = true;         % (default) print status/completion report to screen
taskStructure = essentialTasks;  % metabolic task structure (used instead &quot;taskFile&quot;)
params = [];                % additional optimization parameters for the INIT algorithm
paramsFT = [];              % additional optimization parameters for the fit-tasks algorithm
</code></pre>
<p>Run the getINITModel2 function to extract the GEM</p>
<pre><code class="language-matlab">liverGEM = getINITModel2(refModel, tissue, celltype, hpaData, arrayData, metabolomicsData, removeGenes, taskFile, useScoresForTasks, printReport, taskStructure, params, paramsFT);

% ***Generating model for: liver
% -Using array data
% -Using metabolic tasks
% 
% Reference model statistics:
%         13070 reactions, 3067 genes
%         Mean reaction score: 4.0838
%         Mean gene score: 5.5685
%         Reactions with positive scores: 54.0857%
% 
% Pruned model statistics:
%         11884 reactions, 2494 genes
%         Mean reaction score: 4.1082
%         Mean gene score: 5.351
%         Reactions with positive scores: 54.1905%
% 
% Reactions essential for tasks:
%         386 reactions, 401 genes
%         Mean reaction score: 8.1834
%         Mean gene score: 7.2771
%         Reactions with positive scores: 86.7876%
% 
% MILP detected.
% Set parameter Username
% Set parameter TimeLimit to value 1000
% Set parameter IntFeasTol to value 1e-09
% Set parameter MIPGap to value 1e-12
...
</code></pre>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This algorithm will take a while to run (30 min to 1 hr). It is therefore recommended that this is run on a compute cluster, especially when repeating for many different GEMs.
Alternatively, we recommend using the new "fast" tINIT (<a href="https://github.com/SysBioChalmers/RAVEN/blob/main/INIT/ftINIT.m">ftINIT</a>) algorithm, available from the RAVEN Toolbox. View instructions on using ftINIT <a href="../gem_extraction/">here</a>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In this example, we called <code>getINITModel2</code> with one output, <code>liverGEM</code>. Use <code>help getINITModel2</code> to see a description of the additional optional outputs.</p>
</div>
<p>Once the optimization is complete, we can take a look at the extracted GEM.</p>
<pre><code class="language-matlab">liverGEM

% liverGEM = 
% 
%   struct with fields:
% 
%                      id: 'INITModel'
%             description: 'Automatically generated model for liver'
%                    rxns: {7970×1 cell}
%                    mets: {6740×1 cell}
%                       S: [6740×7970 double]
%                      lb: [7970×1 double]
%                      ub: [7970×1 double]
%                     rev: [7970×1 double]
%                       c: [7970×1 double]
%                       b: [6740×2 double]
%                   comps: {10×1 cell}
%               compNames: {10×1 cell}
%                rxnNames: {7970×1 cell}
%                 grRules: {7970×1 cell}
%              rxnGeneMat: [7970×1904 double]
%              subSystems: {7970×1 cell}
%                 eccodes: {7970×1 cell}
%                rxnNotes: {7970×1 cell}
%                   genes: {1904×1 cell}
%                metNames: {6740×1 cell}
%                metComps: [6740×1 double]
%                  inchis: {6740×1 cell}
%             metFormulas: {6740×1 cell}
%           rxnReferences: {7970×1 cell}
%                 rxnFrom: {7970×1 cell}
%                 metFrom: {6740×1 cell}
%     rxnConfidenceScores: [7970×1 double]
%              metCharges: [6740×1 int64]
%                 version: '1.12.0'
%              annotation: [1×1 struct]
%           unconstrained: [6740×1 double]
%                    name: ''
%                geneFrom: {1904×1 cell}
</code></pre>
<p>It is recommended to change the model <code>id</code> to a more descriptive name than the default of "INITModel". This is particularly useful when analyzing several models together.</p>
<pre><code class="language-matlab">liverGEM.id = 'liver';
</code></pre>
<p>Notice that the model has fewer reactions, metabolites, and genes compared to the reference model.</p>
<pre><code class="language-matlab">numel(ihuman.rxns) - numel(liverGEM.rxns)

% ans =
% 
%         5100

numel(ihuman.mets) - numel(liverGEM.mets)

% ans =
% 
%         3294

numel(ihuman.genes) - numel(liverGEM.genes)

% ans =
% 
%         1163
</code></pre>
<p>It is also good to verify that the extracted GEM can also perform the required metabolic tasks.</p>
<pre><code class="language-matlab">checkTasks(liverGEM, [], true, false, false, essentialTasks);
</code></pre>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.ce72ebac.min.js"></script>
      
    
  </body>
</html>